ntoskrnl/mm/ARM3/miarm.h

```diff
@@ -2339,18 +2339,40 @@ MiDeletePte(
 
 _Must_inspect_result_
 BOOLEAN
-MiMakeSystemAddressValid(
+_MiMakeSystemAddressValid(
     _In_ PVOID PageTableVirtualAddress,
     _In_ PEPROCESS CurrentProcess
 );
 
-ULONG
-NTAPI
-MiMakeSystemAddressValidPfn(
-    IN PVOID VirtualAddress,
-    IN KIRQL OldIrql
+_Must_inspect_result_
+FORCEINLINE
+BOOLEAN
+MiMakeSystemAddressValid(
+    _In_ PVOID PageTableVirtualAddress,
+    _In_ PEPROCESS CurrentProcess)
+{
+    BOOLEAN ret = _MiMakeSystemAddressValid(PageTableVirtualAddress, CurrentProcess);
+    KeMemoryBarrierWithoutFence();
```

Comment:
Why would this be needed? A function call is a sequence point and everything (as visible to the current thread) is evaluated before the return. If the concern is, that this function is inlined on release builds and the write to the PTE is reordered to after access to the page, then we should rather add a memory barrier to the `` MI_WRITE_VALID_PTE``  and ``MI_UPDATE_VALID_PTE`` functions.

Reply:
Correction: In standard C, there is a sequence point between the return and the start of the next full expression, so `` MiMakeSystemAddressValid(PageTableVirtualAddress, CurrentProcess) + *(PUCHAR)PageTableVirtualAddress``  wouldn't have a sequence point, but hopefully nobody would write that code :).

Reply:
It's not about "being evaluated before the return", it's about compiler optimizing memory accesses before or after the call to such functions. Only volatiles are guaranteed to be read or written between sequence points.
I believe that an optimizing compiler can make the following actually access *PointerPte before the call to MiMakeSystemAddressValid :
```
PMMPTE PointerPte = MiAddressToPte(addr);
MiMakeSystemAddressValid(DifferentPointerPteWithinPDRange);
MMPTE TempPte =*PointerPte;
```

Reply:
In that case all code that maps any PTE is prone to this. Therefore I suggest adding the memory barrier to the end of MI_WRITE_VALID_P*E and MI_UPDATE_VALID_P*E, as well as to the beginning of MI_WRITE_INVALID_PTE.

Reply:
LTO may inline all the functions and start reordering code which once was in different ones.
Two questions: why not putting the barrier inside the function, and shouldn't here be something more hard than a compiler barrier? If that's important, how can we be sure CPU won't reorder stuff?

Reply:
```
why not putting the barrier inside the function ?
```
The compiler won't see it on caller side if it's inside the function / in a different compilation unit.
```
If that's important, how can we be sure CPU won't reorder stuff?
```
We can't. But the CPU won't trigger a PF until it's "sure" it has to, and we would have the instructions actually making the PT valid in the pipeline.

Reply:
The CPU doesn't reorder reads and writes on the same CPU core/thread (exception is speculative execution and that doesn't cause faults).
If it's in a different compilation unit and not inlined, then the compiler cannot move memory accesses around it, because it doesn't know the side effects of that function. Otherwise no locking function would ever work without an explicit memory barrier after it.
So the only reason to add that memory barrier is to prevent the compiler from shuffling things around, when it inlines the function and (wrongly) determines that there is no dependency between them . So adding the memory barrier in the function does exactly what is needed.

Reply:
> The CPU doesn't reorder reads and writes on the same CPU core/thread

As far as I understand, this is true only on x86. Not sure if that's important for us right now though